src/cargo/core/compiler/custom_build.rs

```diff
@@ -562,7 +577,39 @@ impl BuildOutput {
                     }
                 }
                 "rustc-cfg" => cfgs.push(value.to_string()),
-                "rustc-env" => env.push(BuildOutput::parse_rustc_env(&value, &whence)?),
+                "rustc-env" => {
+                    let (key, val) = BuildOutput::parse_rustc_env(&value, &whence)?;
+                    // Build scripts aren't allowed to set RUSTC_BOOTSTRAP.
+                    // See https://github.com/rust-lang/cargo/issues/7088.
+                    if key == "RUSTC_BOOTSTRAP" {
+                        // If RUSTC_BOOTSTRAP is already set, the user of Cargo knows about
+                        // bootstrap and still wants to override the channel. Give them a way to do
+                        // so, but still emit a warning that the current crate shouldn't be trying
+                        // to set RUSTC_BOOTSTRAP.
+                        // If this is a nightly build, setting RUSTC_BOOTSTRAP wouldn't affect the
+                        // behavior, so still only give a warning.
+                        if nightly_features_allowed {
+                            warnings.push(format!("Cannot set `RUSTC_BOOTSTRAP={}` from {}.\n\
+                                note: Crates cannot set `RUSTC_BOOTSTRAP` themselves, as doing so would subvert the stability guarantees of Rust for your project.\n\
+                                help: See https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-env for details.",
```

Comment:
I'm a little uncertain, this link doesn't really provide any details about setting `RUSTC_BOOTSTRAP`.  Can you clarify what is meant here?

Reply:
This is probably not a very good link, I don't remember why I added it. Do you think I should link https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html#complications-of-bootstrapping instead, or just remove the link altogether? The dev-guide doesn't talk much about how this affects stability guarantees.

Reply:
I would suggest keeping the link, and adding a commit that updates that section of the Cargo reference to say something like this:

"Note that attempting to set variables that affect the Rust compiler itself is not allowed; for instance, a build script may not set `RUSTC_BOOTSTRAP`."

Reply:
> "Note that attempting to set variables that affect the Rust compiler itself is not allowed; for instance, a build script may not set RUSTC_BOOTSTRAP."

That doesn't seem right though - we allow setting RUSTFLAGS and RUSTDOCFLAGS (and I imagine we'd break half the ecosystem if we prevented that). It's specifically RUSTC_BOOTSTRAP that's disallowed, because it breaks the stability model when any crate can decide on its *own* that it wants to be using unstable now.

Reply:
Probably the place with the most discussion is https://github.com/rust-lang/cargo/issues/7088, but that's a very long and *heated* place for us to point people seeing this for the first time.

Reply:
Let's just have it be a self-contained message then; it seems sufficiently explanatory.

Reply:
Sure thing, done.